### PR TITLE
Allow `PackageSpec` to accept an `AbstractString` for `uuid`

### LINKED ISSUE #2587: Breaking change to `PackageSpec`
On Julia Nightly /  `master`, there appears to be a breaking change to `PackageSpec` (https://github.com/invenia/PkgTemplates.jl/issues/294).

the impacted lines are 
(https://github.com/invenia/PkgTemplates.jl/blob/4302ecb0a8f3304ca519796bae70c196794692c9/src/plugins/tests.jl#L1-L2)
```julia
const TEST_UUID = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
const TEST_DEP = PackageSpec(; name="Test", uuid=TEST_UUID)
```
which now fails with
```julia
ERROR: LoadError: MethodError: Cannot `convert` an object of type String to an object of type Base.UUID
Closest candidates are:
  convert(::Type{T}, ::T) where T at essentials.jl:228
  Base.UUID(::AbstractString) at uuid.jl:82
  Base.UUID(::Any) at uuid.jl:9
Stacktrace:
  [1] convert(#unused#::Type{Union{Nothing, Base.UUID}}, x::String)
    @ Base ./some.jl:36
  [2] Pkg.Types.PackageSpec(name::String, uuid::String, version::Pkg.Versions.VersionSpec, tree_hash::Nothing, repo::Pkg.Types.GitRepo, path::Nothing, pinned::Bool, url::Nothing, rev::Nothing, subdir::Nothing)
    @ Pkg.Types ~/dev/repos/JuliaLang/julia/usr/share/julia/stdlib/v1.7/Pkg/src/Types.jl:92
  [3] Pkg.Types.PackageSpec(; name::String, uuid::String, version::Pkg.Versions.VersionSpec, tree_hash::Nothing, repo::Pkg.Types.GitRepo, path::Nothing, pinned::Bool, url::Nothing, rev::Nothing, subdir::Nothing)
    @ Pkg.Types ./util.jl:471
  [4] top-level scope
    @ ~/.julia/packages/PkgTemplates/1TTni/src/plugins/tests.jl:2
  [5] include(mod::Module, _path::String)
    @ Base ./Base.jl:389
  [6] include(x::String)
    @ PkgTemplates ~/.julia/packages/PkgTemplates/1TTni/src/PkgTemplates.jl:2
```
and instead now requires the `uuid` keyword to be passed a `UUID` rather than a `String`.
(https://github.com/invenia/PkgTemplates.jl/pull/295)

The PR introducing the breaking change seems to be #2477 (https://github.com/JuliaLang/Pkg.jl/pull/2477#issuecomment-852892246)

### PR DESCRIPTION
**Pull request by DilumAluthge**
Fixes #2587 

With this PR: 

```julia
julia> import Pkg

julia> x = Pkg.PackageSpec(; name = "Example", uuid = "7876af07-990d-54b4-ab0e-23690620f79a")
PackageSpec(
  name = Example
  uuid = 7876af07-990d-54b4-ab0e-23690620f79a
  version = *
)

julia> typeof(x.uuid)
Base.UUID
```

### CODE REVIEW

File: src/Types.jl [hunk 1]
```diff
@@ -88,18 +88,46 @@ Base.:(==)(r1::GitRepo, r2::GitRepo) =
...
+    path::Union{Nothing,String}
+    pinned::Bool
     # used for input only
-    url = nothing
-    rev = nothing
-    subdir = nothing
+    url
+    rev
+    subdir
+    function PackageSpec(; name::Union{Nothing,String} = nothing,
```
Comment (fredrikekre) [inline src/Types.jl:103]: This will now remove the default constructors so perhaps this should be an outer one?

File: src/Types.jl [hunk 2]
```diff
@@ -88,18 +88,46 @@ Base.:(==)(r1::GitRepo, r2::GitRepo) =
...
+                           path::Union{Nothing,String} = nothing,
+                           pinned::Bool = false,
+                           url = nothing,
+                           rev = nothing,
+                           subdir = nothing)
+        if uuid isa AbstractString
+            _uuid = Base.UUID(uuid)
+        else
+            _uuid = uuid
+        end
```
Comment (fredrikekre) [inline src/Types.jl:117]:
```suggestion
        uuid = uuid === nothing ? nothing : UUID(uuid)
```

File: src/Types.jl [hunk 3]
```diff
@@ -88,18 +88,46 @@ Base.:(==)(r1::GitRepo, r2::GitRepo) =
...
+            _uuid,
+            version,
+            tree_hash,
+            repo,
+            path,
+            pinned,
+            url,
+            rev,
+            subdir,
+        )
```
Comment (fredrikekre) [inline src/Types.jl]:
```suggestion
        return new(name, uuid, version, tree_hash, repo, path, pinned, url, rev, subdir)
```

Comment (DilumAluthge) [reply src/Types.jl]: GitHub won't let me "apply" this suggestion directly, but I have made this change manually.

File: src/Types.jl (hunk 1 above)
Comment (DilumAluthge) [reply src/Types.jl]: Good call, yeah I've moved it to be an outer constructor now.

**Comment by DilumAluthge** [comment]
I think this is good to go now?

**Comment by fredrikekre** [comment]
Could have added tests

**Comment by DilumAluthge** [comment]
Definitely. Tests in https://github.com/JuliaLang/Pkg.jl/pull/2593

**Pull request merged**